### DISCUSSION TITLE
HTML API: Possibly-interesting patches.

### DISCUSSION THREAD

**Discussion by dmsnell** [WordPress/gutenberg#58755] [HTML API]
[HTML issues](https://github.com/WordPress/gutenberg/discussions/57560) | [Refactors](https://github.com/WordPress/gutenberg/discussions/54750) | [Interesting Patches](https://github.com/WordPress/gutenberg/discussions/58755) | [Broader Roadmap](https://github.com/WordPress/gutenberg/discussions/54583)

### Testing code
 - https://github.com/WordPress/wordpress-develop/pull/6044 Add this to run HTML API tests locally without any complicated setup.
 - https://github.com/WordPress/wordpress-develop/pull/6050 Generate coverage reports for the HTML API.
 - https://github.com/WordPress/wordpress-develop/pull/6121 Adds debug methods to pretty-print internal state to the console.

### Benchmarking code
 - https://github.com/WordPress/wordpress-develop/pull/6049 Compare runtime performance of HTML Processor against Tag Processor

### Tools built with the HTML API

 - https://github.com/WordPress/wordpress-develop/pull/6045 HTML syntax-highlighter and formatter.
 - https://github.com/WordPress/wordpress-develop/pull/6046 Print a diagnostics view of an HTML document, listing every token in order.
 - https://github.com/WordPress/wordpress-develop/pull/6051 Print a diagnostics view of an HTML document, showing breadcrumb structure and attribute preview, and whether the document is fully-supported.
 - https://github.com/WordPress/wordpress-develop/pull/6047 Render HTML in a terminal-friendly way.
 - https://github.com/WordPress/wordpress-develop/pull/6048 Search HTML with a regex, show containing HTML for results.

### Other

 - https://github.com/WordPress/wordpress-develop/pull/6052 Rough JS port of the Tag Processor for basic performance comparisons of PHP against "something that should be faster"
 - https://gist.github.com/dmsnell/11db00126a1c039cc5cb4f34f0d83ef4 Rough C port of the Tag Processor for performance estimation of "how fast could this reasonably be if included inside PHP?"

**Comment by westonruter** [comment]
**Image Loading Optimization** (https://github.com/WordPress/performance/issues/869) with a [tag processor](https://github.com/WordPress/performance/blob/feature/image-loading-optimization/modules/images/image-loading-optimization/class-ilo-html-tag-processor.php) which walks over the document which can compute the XPath of each tag it visits.
